### ISSUE TITLE
Thank YOU!

### ISSUE THREAD

**Issue by brothermechanic** [waebbl/waebbl-gentoo#4]
Hello my friend. Thank you for working freecad with deps! All works great! I see that you used my stuff, i really can't to finish it; because i dont developer, i designer)
Now i find your overlay, and now i user of waebbl-gentoo! 
Thamk you!
![screenshot_20180905_132113](https://user-images.githubusercontent.com/3109995/45087593-00ffaf00-b0f6-11e8-85a2-b96308ea7bbf.jpg)

**Comment by brothermechanic** [comment]
Oh, you need add sci-libs/libmed to waebbl repo

**Comment by brothermechanic** [comment]
small comment;
can you remove
EGIT_BRANCH="master" 
from freecad-9999.ebuild
?
because i can't to select git branch externally from portage

**Comment by waebbl** [comment]
Hello brothermechanic, glad to hear, you like the ebuild, but tbh, without your ebuild as a base, it would have been much more work. Thanks a lot for the good work.
I was thinking whether we could work together on some of your cg ebuilds. It's a themed repository name, so I think much more people will choose it when they are looking for cg related ebuilds. You could add me as a contributor for the repo, so I could directly push into it. Of course, if you don't like this idea, I can also use PR's for this. On the freecad ebuild I was working together with a friend, @rigid, to my knowledge he already sent you a PR with an earlier state of the ebuild a couple of days ago, but this PR would be outdated now.

> Oh, you need add sci-libs/libmed to waebbl repo

I was using the libmed from your repository and found it's working well, so I didn't see the need to add it to my repository too. Why do you think I should add it?

Does the master branch did not work for you? I chose it, because I found the releases/freecad-0.17 branch was not compiling easily, so I tested the master branch, which seems to be more robust currently. I first need to check whether it compiles without any EGIT_BRANCH variable. If so, I might remove it. You can also override this setting from the environment by setting ${PN}_LIVE_BRANCH to the desired branch when calling emerge. So using `FREECAD_LIVE_BRANCH="whathever-branch-you-like" emerge =freecad-9999` should do the work. Take a look at /usr/portage/eclass/git-r3.eclass and look for BRANCH, where this is explained.

**Comment by brothermechanic** [comment]
Hello
Yes of course, i want to work together!
Because as i said i'm designer and codding part is hard to me. And i do it because i haven't choise!
I sent invition to you!

**Comment by brothermechanic** [comment]
libmed must be together with freecad, i think

**Comment by brothermechanic** [comment]
I found bug. I have freecad in appimage (because i cant install it normally, and it work fine. Now i install your freecad and it work great.
BUT
When i reboot computer i get very slow boolean (union/diffence etc) in freecad, both versions (appimage and your)!!!!!???? What is it? Boolean operations now ~ 7 times slower than before your freecad instalation!
I remove freecad and its depend packages, no result.
Then i rebooted. And now my freecad by appimage worked fine agane!
I found that the problem is in opencascade package
[test.zip](https://github.com/waebbl/waebbl-gentoo/files/2358425/test.zip)
This is my test file. To reply error do
1 - on clean system download official freecad appimage
2 - test to boolean my 2 objects and remember time to operation ( must be 1 second or less)
3 - install your freecad and do step 2  ( must be 1 second or less)
4 - rebot computer
5 do step 2  ( must be 7 second or more)
This is serious problem, because on complex models this time is 5 minutes!
Thank you
Hope you solve this problem!

**Comment by waebbl** [comment]
Thanks for the invitation. I won't add any new ebuilds before talking to you about it.

About your problem. I doubt I can be of much help with this. I can take a look into it, but I'm new to freecad, never used it before and just wanted to provide a useable ebuild, when it was removed from the official tree.

I agree with you that it seems to be related to opencascade. The related source code file should be src/Mod/Part/App/FeaturePartBoolean.cpp where the boolean operations are done using occ algorithms. When I strace'd the operation it looked like it's waiting for some threads or mutexes to be released to continue. I compiled occ with tbb use flag enabled.  You might try switching the tbb use flag on opencascade and recompile whether this changes anything.

What's strange about it, is that the appimage also get's slower. It usually has almost all dependant libraries included and use only the most common libraries from the system it is running, i.e. only glibc and probably some X and the opengl library will be used. So the appimage will most likely not use the opencascade installed on the system, but it's own opencascade, although I'm not sure about this. Additionally, the appimage is version 0.17 and the ebuild is a live ebuild with version 0.18, where a lot has changed since. That's why I'm thinking it might have another source than the ebuild.

You should go to the freecad forum and asked there, whether they have any idea of what's the reason for this slow operation.

**Comment by brothermechanic** [comment]
Yes, you understand me right!
I try -tbb.
Very strange issue....

**Comment by waebbl** [comment]
On my machine it's also happening without a reboot. I only tried this with my currently installed freecad and the appimage version, both with the same result, of taking several seconds until the operation is finished.

I also checked the appimage again, it's statically linked, i.e. it does not even rely on the system's glibc or any other lib, but has all needed libraries included.

From this, I would conclude, it might have something to do with the kernel. So, with freshly booted kernel it works well, but when the kernel runs for some time, the delay is starting. Have you tried the appimage, after the machine has been running for some time? Is it still fast with the operation, or does it start to take more time?

Have you already asked on the freecad forum about this?

**Comment by brothermechanic** [comment]
Thank you for testing!
Can we talking directly by facebook, email, matrix etc in future?
(I prefer matrix, this is opensource chat.)
No, i dont write on freecad forum, this is hard to me, because my english is very weak.
"Appimage has all needed libraries included" - yes, it's strange, my other programs affect ti appimage???

reply; 
1) when i remove opencascade package, and obligatorily REBOT computer 
appimage's boolean work as it how should work.
2) when i install opencascade and NOT reboot computer 
appimage's boolean work as it how should work.
and your freecad work perfect (better than appimage)

and i work without reboot long time (several days with suspends)
problem comes when i

1) first reboot computer with opencascade installed
2) reboot computer and remove opencascade package (problem gone only after reboot)

I have systemd installed (do you too?)
May be problem is in some process, or clearing cache or in /etc/env.d/opencascade ?

Can you add 7.2 or another version of opencascade?
(opencascade compiles take time very long time, can we optimise package size?)

ps: i test build with vtk-7* and vtk-8* - same for me)
And please add EGIT_BRANCH="master" to freecad )))

**Comment by waebbl** [comment]
I'm currently on the way to stop using facebook, so this doesn't make sense for me. We can use matrix, I have a client installed on my phone and I'm looking to get the quaternion matrix client into my overlay.

I'm also using systemd, but I use only vtk-8 packages. I always had some trouble compiling the vtk-7 packages on gentoo. I also haven't tried to reboot, because my computer is usually always running and only suspended. I only reboot if it's necessary, i.e. after kernel upgrades or something. I also don't want to remove and reinstall large packages like vtk, opencascade or freecad all the time, because the compile times are quite long on my machine.
I can try to add opencascade-7.2.0p3, but this will take some time. I don't want to add any older versions, especially no 6.x version, because they seem not be available on their public git repository and might no longer be supported by the occt team.

IMO your english is good enough for the forum. They don't care much about english skills, because they have people from all over the world contributing to the forum and to freecad. They even have a russian subforum, so can ask in your native language there first.

**Comment by Perlovka** [comment]
@brothermechanic
Try to cleanup FreeCAD files from /tmp. There are some files and directories that FreeCAD creates while running, don't sure appimage version or both.
/tmp directory usually being cleaned upon reboot, so this might be a problem.

**Comment by Perlovka** [comment]
Very likely this bug with boolean operations caused by CSF_DEBUG_BOP is being set in /etc/env.d/opencascade/7.3.0

https://dev.opencascade.org/doc/overview/html/occt_dev_guides__debug.html

**Comment by brothermechanic** [comment]
Privet Perlovka, ofcourse /tmp dir is clear, because i reboot my computer, do you read above?

**Comment by waebbl** [comment]
I already tested cleaning ${TMP} but it doesn't help with this.
It's indeed because of the CSF_DEBUG_BOP variable from the environment, just verified it.
If you use `unset CSF_DEBUG_BOP` in your terminal and then run freecad, the operation get's as fast as one would expect.
For a workaround comment the line containing CSF_DEBUG_BOP in /etc/env.d/51opencascade. I will fix this in the repository.
Thanks @Perlovka for the hint!

**Comment by Perlovka** [comment]
@waebbl I've just commited PR to gentoo main tree with opencascade version bump. Removed some things, because they need some more testing. If this PR will be merged, we could improve it further. :wink:

**Comment by rigid** [comment]
@Perlovka i doubt the current state would be ready to be merged into the main tree just yet.
For example, it use tbb optimized memory management (MMGT_OPT=2) regardless of the tbb USE flag setting.

**Comment by Perlovka** [comment]
I've set +tbb as it was in 6.9.1 ebuild, so this should not be a problem for a default install. I hope )

**Comment by rigid** [comment]
ah, then that shouldn't be an issue. I also see that inspector.sh hasn't it's world-writable flag removed. I'm not sure, will -DBUILD_Inspector be true when it's detected?
Besides that, I don't see anything major...

**Comment by Perlovka** [comment]
I decided not to install inspector because AFAIR there was issue with a paths in script, also I dropped tests and qt5 for now. I have no much time to test every feature, so I created basic ebuild to be able to build other apps with newer version.  One could make another PR with extended features after this version gets into a portage )

**Comment by rigid** [comment]
right, maybe @waebbl could do that after some more tests?

**Comment by brothermechanic** [comment]
Well, here I am)
My tests:
1) I commit CSF_DEBUG_BOP in /etc/env.d/opencascade/7.3.0 (/etc/env.d/51opencascade is just symlink) - don't help
2) remove /etc/env.d/opencascade dir and symlink /etc/env.d/51opencascade - helped! after reboot i get worked booleans in freecad from appimage, but in freecad from waebbl i get this
![screenshot_20180916_174913](https://user-images.githubusercontent.com/3109995/45597708-b3215b80-b9bf-11e8-89b9-472abd85a6fd.jpg)
(i think that freecad lost opencadcade link)
3) For fixing  freecad from waebbl i 
`cp /var/lib/layman/waebbl/sci-libs/opencascade/files/51opencascade  /etc/env.d/`
Does not help.
I think that the problem is somewhere in env

**Comment by brothermechanic** [comment]
Do we really need eselect-opencascade package?

**Comment by rigid** [comment]
please retry with the current ebuild from today:
```
emerge -1va opencascade
# this or logout/login
env-update
source /etc/profile
freecad
```

**Comment by waebbl** [comment]
You don't need to remove the file, just comment the one line in this file. It doesn't matter if you change it in /etc/env.d/51opencascade or in /etc/env.d/opencascade/7.3.0 because the former is just a symlink to the latter which is managed by eselect-opencascade.

For now we need eselect-opencascade as the official tree contains opencascade-6.9.1-r1 in a different slot, which probably has completely different env settings. As soon as official tree has added the PR from @Perlovka _**and**_ tree-cleaned the 6.9.1 slot, it might no longer be necessary. Tree-cleaning will eventually take quite some time (think of months).

If you already commented the line, reread my earlier post about this topic. You also need to unset the variable in any active shell you want to use with freecad or any other app that uses opencascade!

**Comment by brothermechanic** [comment]
Hello, friends
Now opencascade works perfectly! Freecad works perfectly!
Thank you all!
